### PR TITLE
Add speedtest

### PR DESCRIPTION
**Pull request by tharvik**
Finally done, that's the last page for the demo :1st_place_medal:

* un-mock speedtest
  * with progress bars!
  * compute TPS on the fly
  * refresh page every few transactions (not for each one)
  * send a bunch before refreshing
* add agent/get_last_sequence

### CODE REVIEW

File: at2-ns/src/user/full.rs [hunk 1]
```diff
@@ -23,4 +25,9 @@ impl FullUser {
     pub fn keypair(&self) -> &sign::KeyPair {
         &self.keypair
     }
+
+    /// Return a thin version of the user
```
Comment (ineiti) [inline at2-ns/src/user/full.rs:29]: Is this a good way to lose some pounds?

Can you describe in just one phrase what the difference between a `thin` and a `fat` user is?

File: web/src/agents/node/get_last_sequence.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
+use at2_node::client::{self, Client};
+use at2_ns::ThinUser;
+use wasm_bindgen_futures::spawn_local;
+use yew::worker::*;
+
+use crate::config::Config;
+
+/// Get the last sequence for a user
```
Comment (ineiti) [inline web/src/agents/node/get_last_sequence.rs]: Sequence of what? Transactions?

File: web/src/agents/node/get_last_sequence.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    client: Client,
+}
+
+impl Agent for GetLastSequence {
+    type Reach = Context<Self>;
+    type Message = (HandlerId, Result<sieve::Sequence, client::Error>);
+    type Input = ThinUser;
+    type Output = Result<sieve::Sequence, client::Error>;
+
+    fn create(link: AgentLink<Self>) -> Self {
```
Comment (ineiti) [inline web/src/agents/node/get_last_sequence.rs:24]: FoMyRuK: why is it `create` and not `new`?

File: web/src/agents/node/get_last_sequence.rs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    fn update(&mut self, (id, ret): Self::Message) {
+        self.link.respond(id, ret);
+    }
+
+    fn handle_input(&mut self, user: Self::Input, id: HandlerId) {
+        let mut client = self.client.clone();
+        let callback = self.link.callback(|ret| ret);
+
+        spawn_local(async move {
+            callback.emit((id, client.get_last_sequence(user.public_key()).await));
```
Comment (ineiti) [inline web/src/agents/node/get_last_sequence.rs:42]: Why does `get_last_sequence` modify the client?

File: web/src/pages/summary.rs [hunk 5]
```diff
@@ -22,7 +22,21 @@ impl Component for Summary {
         html! { <>
             <h1> { "Summary" } </h1>
 
-            <p> { "We showed that AT2 is faster than bitcoin." } </p>
+            <p> { "We showed that AT2 is faster than other well known
```
Comment (ineiti) [inline web/src/pages/summary.rs:25]: As discussed, please start a google-doc and invite me and Christian to edit it. This should be much longer and explain more in-depth what is happening. It should also link directly to the paper and perhaps other resources.

**Review by ineiti — CHANGES_REQUESTED** [review]
Some questions and requests for augmenting of comments.

File: at2-ns/src/user/full.rs (hunk 1 above)
Comment (tharvik) [reply at2-ns/src/user/full.rs:29]:
```rust
//! There is two types of user,
//! one having a KeyPair which is usually the current user,
//! and one having only a PublicKey, representing the other users on the network.
```

Both are "user", but if you find better names, go for it, I'm not really proud of it.

File: web/src/agents/node/get_last_sequence.rs (hunk 2 above)
Comment (tharvik) [reply web/src/agents/node/get_last_sequence.rs]: every message send on the network is a tuple of the sender, it's counter/sequence number, and a payload. this agent retrieve the latest processed sequence number for a given user, to show the number of confirmed transactions during a speedtest.

And now that I'm writing it, it's indeed common knowledge just for me; I'm adding it to the comment.

File: web/src/agents/node/get_last_sequence.rs (hunk 3 above)
Comment (tharvik) [reply web/src/agents/node/get_last_sequence.rs:24]: there isn't constructor per say in Rust; function returning `Self` can be used to construct object but there aren't default. `new` is the common way to name it.
This `create` function is in the `Agent` trait of "yew", and goes with its semantic (there is a `destroy` function, components can get recreated on page change, ...)

File: web/src/agents/node/get_last_sequence.rs (hunk 4 above)
Comment (tharvik) [reply web/src/agents/node/get_last_sequence.rs:42]: It uses a TCP stream internally and needs to modify it (incrementing pkg count, reading and writing to the stream, ...), so the client itself will need to mutate state to do so.

btw, as Rust disallow having double mutable reference, we need to `clone` the client for each new input, as the previous running `spawn_local` may not be finished at this point.

File: web/src/pages/summary.rs (hunk 5 above)
Comment (tharvik) [reply web/src/pages/summary.rs:25]: Shared in engineers' chat

File: at2-ns/src/user/full.rs (hunk 1 above)
Comment (ineiti) [reply at2-ns/src/user/full.rs:29]: `There are two types of users`

I think I called them `User` and `Contact`. The first has the private key, the second doesn't.

**Review by ineiti — APPROVED** [review]

File: at2-ns/src/user/full.rs (hunk 1 above)
Comment (tharvik) [reply at2-ns/src/user/full.rs:29]: Haa, well played, moving to `User` & `Contact`, thanks :)

**Pull request merged**